packages/yarnpkg-pnpify/sources/hoist.ts

```diff
@@ -0,0 +1,283 @@
+ */
+export interface HoisterPackageInfo {
+  /** The name of the package */
+  name: HoisterPackageName;
+  /** The own weight of the package, without its transitive dependencies */
+  weight: HoisterWeight;
+}
+
+/**
+ * The results of weighting each package with its transitive dependencies in some subtree.
+ */
+type WeightMap = Map<HoisterPackageId, HoisterWeight>;
+type ReadonlyWeightMap = ReadonlyMap<HoisterPackageId, HoisterWeight>;
+
+/**
+ * Hoists package tree, by applying hoisting algorithm to each tree node independently.
+ * We first try to hoist all the packages from anywhere in the whole tree to the root package.
+ * Then we apply the same algorithm to the subtree that starts at one of the dependencies of the
+ * root package, we do this for each dependency, then move further down to dependencies of
+ * dependencies, etc.
+ *
+ * This function does not mutate its arguments, it hoists and returns tree copy
+ *
+ * @param tree package tree (cycles in the tree are allowed)
+ * @param packageInfos package infos
+ * @param nohoist package ids that should be excluded from applying hoisting algorithm. Nohoist
+ *                packages can be hoisted themselves, and their dependencies can be hoisted too,
+ *                but only to the package itself, they cannot be hoisted to the nohoist package parent
+ *
+ * @returns hoisted tree copy
+ */
+export const hoist = (tree: ReadonlyHoisterPackageTree, packageInfos: ReadonlyArray<HoisterPackageInfo>, nohoist: ReadonlySet<HoisterPackageId> = new Set()): HoistedTree => {
+  // Make tree copy, which will be mutated by hoisting algorithm
+  const treeCopy: HoisterPackageTree = tree.map(({deps, peerDeps}) => ({deps: new Set(deps), peerDeps: new Set(peerDeps)}));
+
+  const seenIds = new Set<HoisterPackageId>();
+
+  const hoistSubTree = (nodeId: HoisterPackageId) => {
+    seenIds.add(nodeId);
+
+    // Apply mutating hoisting algorithm on each tree node starting from the root
+    hoistInplace(treeCopy, nodeId, packageInfos, nohoist);
+
+    for (const depId of treeCopy[nodeId].deps) {
+      if (!seenIds.has(depId)) {
+        hoistSubTree(depId);
+      }
+    }
+  };
+
+  if (treeCopy.length > 0 && treeCopy[0].deps.size > 0)
+    hoistSubTree(0);
+
+  return treeCopy.map(({deps}) => deps);
+};
+
+/**
+ * Performs package subtree hoisting to its root.
+ * This funtion mutates tree.
+ *
+ * @param tree package tree
+ * @param rootId package subtree root package id
+ * @param packages package infos
+ * @param nohoist nohoist package ids
+ */
+const hoistInplace = (tree: HoisterPackageTree, rootId: HoisterPackageId, packages: ReadonlyArray<HoisterPackageInfo>, nohoist: ReadonlySet<HoisterPackageId>): void => {
+  // Get the list of package ids that can and should be hoisted to the subtree root
+  const hoistedDepIds = computeHoistCandidates(tree, rootId, packages, nohoist);
+  const seenIds = new Set<HoisterPackageId>();
+
+  const removeHoistedDeps = (nodeId: HoisterPackageId) => {
```

Comment:
I wonder if we could keep track in `computeHoistCandidates` of which node depends on those who get hoisted, so that we wouldn't have to traverse the all subtree all over again 🤔

Reply:
As discussed on Discord, we actually should do multiple things to compute weight properly:
1. Compute weight only for the packages that has multiple version as candidates for hoisting
2. As you properly noticed, we should keep track of ancestors, because the total weight of a package for hoisting purposes is the weight of the packages plus all its transitive dependencies multiplied by the number of ancestors that are using this package.
3. We should still traverse the all subtree to find out weights of all the transitive dependencies, because out tree always changes during hoisting and at the same time we should keep track of ancestor changes.
So we kinda should do it all :)
I believe I've implemented all 3 points now